src/sys/unix.rs

```diff
@@ -959,21 +956,9 @@ impl crate::Socket {
         if len == 0 {
             Ok(None)
         } else {
-            // Allocate a buffer for `CString` with the length including the
-            // null terminator.
-            let len = len as usize;
-            let mut name = Vec::with_capacity(len);
-
             // TODO: use `MaybeUninit::slice_assume_init_ref` once stable.
-            // Safety: `len` bytes are writen by the OS, this includes a null
-            // terminator. However we don't copy the null terminator because
-            // `CString::from_vec_unchecked` adds its own null terminator.
-            let buf = unsafe { slice::from_raw_parts(buf.as_ptr().cast(), len - 1) };
-            name.extend_from_slice(buf);
-
-            // Safety: the OS initialised the string for us, which shouldn't
-            // include any null bytes.
-            Ok(Some(unsafe { CString::from_vec_unchecked(name) }))
+            let buf = &buf[..len.try_into().unwrap()];
+            Ok(Some(unsafe { mem::transmute::<_, &[u8]>(buf) }.into()))
```

Comment:
Let's not use `transmute`, it's too easy to misuse. Also let's keep the same code as in `MaybeUninit::slice_assume_init_ref` so we can switch to that easily once stable.

It think the following would be shortest.

```rust

// Safety: `len` bytes are writen by the OS, this includes a null
// terminator. However we don't need the null terminator.
let buf = unsafe { slice::from_raw_parts(buf.as_ptr().cast(), (len - 1) as usize) };
Ok(Some(buf.into_vec()))
```

Reply:
`len` does _not_ include a null terminator AFAIK.

Other than that, this code _is_ the same as it would be with `slice_assume_init_ref`. That code would be:
```rust
let buf = &buf[..len.try_into().unwrap()];
Ok(Some(unsafe { MaybeUninit::slice_assume_init_ref(buf) }.into()))
```

Reply:
`len` does include a null terminate, I've just manually tried this patch. The test currently doesn't fail it it doesn't have permissions (mainly for the CI), but we might have to change that. I'll fiddle with this a bit more and merge this changing the signature to `[u8]` rather than `CStr`.

P.S. it requires `sudo setcap cap_net_raw=ep $test_bin`, something that we might have to do in the CI as well.

Reply:
I'm about 95% sure that len *does not* include the null terminator when the name is IFNAMSIZ characters long. That is probably an edge case worth testing.

Reply:
After a bit more manual testing. I couldn't create an interface with a name larger than 15 bytes, with `IFNAMSIZ` being 16.

```bash
sudo ip link add test_interface0 type bridge  # Fine.
sudo ip link add test_interface01 type bridge # Error: Attribute failed policy validation.
```

I also tested binding to `test_interface01` with the implementation in #187 which worked.